### PR TITLE
Ruby 2.7 keyword arguments support (& related deprecations)

### PR DESCRIPTION
**Pull request by thbar**
See Changes.md for details & deprecations.

In particular, `SourceTransformAdapter` has been removed (an upgrade path is provided in Changes.md).

### Related

* https://github.com/thbar/kiba/pull/86
* https://github.com/thbar/kiba/issues/85

### CODE REVIEW

File: lib/kiba-common/transforms/source_transform_adapter.rb [hunk 1]
```diff
@@ -3,7 +3,14 @@ module Common
     module Transforms
       class SourceTransformAdapter
         def process(args)
-          args.shift.new(*args).each do |row|
+          klass = args.shift
```
Comment (thbar) [inline lib/kiba-common/transforms/source_transform_adapter.rb]: `args.shift` is modifying the original array. It would be better to avoid doing so.

**Comment by ioquatix** [comment]
There should be a trivial way to handle the arguments correctly.

I'm happy to dig into this but @eregon may have some more ideas since he worked on this in the past.

**Comment by thbar** [comment]
@ioquatix if you have any suggestions, these are most welcome!

Before contributing, check out [Contributing & Legal](https://github.com/thbar/kiba-common#contributing--legal) (I do not have a proper Contributor Agreement published, although that will come out later) - inspired by what [Sidekiq does](https://github.com/mperham/sidekiq/blob/master/.github/contributing.md).

In the mean time I'm going to achieve further testing with the existing code, but happy to read your (or @eregon's) propositions!

**Comment by eregon** [comment]
I kind of replied in a generic way regarding delegation methods here:
https://discuss.rubyonrails.org/t/new-2-7-3-0-keyword-argument-pain-point/74980/6?u=benoit_daloze

**Comment by thbar** [comment]
@eregon thanks for your reply, appreciated! I think I will wait a bit and see how things evolve in future minor versions of Ruby 2.7.

File: lib/kiba-common/transforms/source_transform_adapter.rb [hunk 2]
```diff
@@ -3,7 +3,14 @@ module Common
...
+          if args.last.is_a?(Hash) && RUBY_VERSION >= "2.7"
+            kwargs = args.pop
+            instance = klass.new(*args, **kwargs)
+          else
+            instance = klass.new(*args)
+          end
```
Comment (eregon) [inline lib/kiba-common/transforms/source_transform_adapter.rb]: For this case and all delegation cases, the only reliable way for current Ruby 2.7.0 & 2.7.1, is to use `ruby2_keywords def process(*args)`. That flags the keyword arguments, so once `*args` is used in the body they'll be keyword arguments again.

If you don't want to change the signature of `process(args)`, you'll need to find a caller which passes them as `*args` and use `ruby2_keywords` there. You can check if the keyword Hash is flagged correctly with `Hash.ruby2_keywords_hash?`.

**Comment by mame** [comment]
Hi!  I'm the ringleader of this incompatibility.  I'm sorry for bothering you, and thank you for giving us the first comment to [matz's listening](https://discuss.rubyonrails.org/t/new-2-7-3-0-keyword-argument-pain-point/74980/1) :-)

Your code that depends upon `RUBY_VERSION <= "2.7"` has a (very subtle) corner case.  The following demonstrates the issue:

```ruby
require 'kiba'
require 'kiba-common/sources/enumerable'
require 'kiba-common/transforms/source_transform_adapter'
require 'kiba-common/destinations/lambda'

class MyClass
  def initialize(r, empty_hash)
    @r = r
  end

  def each
    yield @r
  end
end

job = Kiba.parse do
  source Kiba::Common::Sources::Enumerable, [1, 2, 3]

  transform do |r|
    [MyClass, r, {}]
  end

  transform Kiba::Common::Transforms::SourceTransformAdapter

  destination Kiba::Common::Destinations::Lambda, on_write: -> (r) { p r }
end

Kiba.run(job)

# $ bundle exec ruby -Ilib demo.rb
# /tmp/kiba-common/lib/kiba-common/transforms/source_transform_adapter.rb:9: warning: Passing the keyword argument as the last hash parameter is deprecated
# demo.rb:7: warning: The called method `initialize' is defined here
# 1
# 2
# 3
```

This PR still emits a warning because the last hash of `[MyClass, r, {}]` is passed as `MyClass.new(r, **{})` by [this call](https://github.com/thbar/kiba-common/pull/27/files/55903078abdf6842cc25e523e765de8b01685e7d#diff-19618a2a6558fe98b4aefdaf37f99043R9).  This issue occurs only when passing an empty hash as the last argument.  (If you think it is not a matter, I think your fix is okay.)

This is my fix candidate:

```diff
diff --git a/lib/kiba-common/destinations/csv.rb b/lib/kiba-common/destinations/csv.rb
index 6fc3a09..6aca3e1 100644
--- a/lib/kiba-common/destinations/csv.rb
+++ b/lib/kiba-common/destinations/csv.rb
@@ -13,7 +13,7 @@ module Kiba
         end
         
         def write(row)
-          @csv ||= ::CSV.open(filename, 'wb', csv_options)
+          @csv ||= ::CSV.open(filename, 'wb', **csv_options)
           @headers ||= row.keys
           @headers_written ||= (csv << headers ; true)
           csv << row.fetch_values(*@headers)
diff --git a/lib/kiba-common/sources/csv.rb b/lib/kiba-common/sources/csv.rb
index 2a45a33..3ae6837 100644
--- a/lib/kiba-common/sources/csv.rb
+++ b/lib/kiba-common/sources/csv.rb
@@ -12,7 +12,7 @@ module Kiba
         end
 
         def each
-          ::CSV.foreach(filename, csv_options) do |row|
+          ::CSV.foreach(filename, **csv_options) do |row|
             yield row
           end
         end
diff --git a/test/test_csv_source.rb b/test/test_csv_source.rb
index b99a7d2..6f14485 100644
--- a/test/test_csv_source.rb
+++ b/test/test_csv_source.rb
@@ -18,7 +18,7 @@ class TestCSVSource < MiniTest::Test
     FileUtils.rm(TEST_FILENAME) if File.exist?(TEST_FILENAME)
   end
 
-  def run_job(*args)
+  ruby2_keywords def run_job(*args)
     rows = []
     job = Kiba.parse do
       source Kiba::Common::Sources::CSV, *args
diff --git a/test/test_integration.rb b/test/test_integration.rb
index b93a9d0..f16dc7b 100644
--- a/test/test_integration.rb
+++ b/test/test_integration.rb
@@ -31,8 +31,10 @@ class TestIntegration < Minitest::Test
         transform do |r|
           [
             Kiba::Common::Sources::CSV,
-            filename: r,
-            csv_options: { headers: true, header_converters: :symbol }
+            Hash.ruby2_keywords_hash({
+              filename: r,
+              csv_options: { headers: true, header_converters: :symbol }
+            })
           ]
         end
 
```

and for kiba

```diff
diff --git a/lib/kiba/context.rb b/lib/kiba/context.rb
index 2511739..53c5780 100644
--- a/lib/kiba/context.rb
+++ b/lib/kiba/context.rb
@@ -8,15 +8,15 @@ module Kiba
       @control.pre_processes << { block: block }
     end
 
-    def source(klass, *initialization_params)
+    ruby2_keywords def source(klass, *initialization_params)
       @control.sources << { klass: klass, args: initialization_params }
     end
 
-    def transform(klass = nil, *initialization_params, &block)
+    ruby2_keywords def transform(klass = nil, *initialization_params, &block)
       @control.transforms << { klass: klass, args: initialization_params, block: block }
     end
 
-    def destination(klass, *initialization_params)
+    ruby2_keywords def destination(klass, *initialization_params)
       @control.destinations << { klass: klass, args: initialization_params }
     end
 
```

With this patch, `bundle exec rake test` shows no warnings.

As @eregon said in the forum, a perfect solution is to use `ruby2_keywords`.  I'd like you to read [this article](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/) in detail, but in short, `ruby2_keywords def foo(*a); bar(*a); end` means `def foo(*a); bar(*a); end` in 2.6 and 2.7 (with no wraning), and `def foo(*a, **opt); bar(*a, **opt); end` in 3.0 or later.  The trick is that a hash object has a dedicated internal flag when it is passed as keywords.

One very unfortunate thing is that your DSL handles `[MyCSVSource, filename: r]` as `MyCSVSource.new(filename: r)`.  The array literal is currently interpreted as `[MyCSVSource, { filename: r }]`, so the last hash is not handled as keywords.  I couldn't predict this usage of an array literal, so there is no simple solution.  In my patch, `Hash.ruby2_keywords_hash(...)` is used to explicitly mark the hash object as keywords, but I think it is not acceptable for your DSL.

My proposal is to avoid this usage of array literals, and to provide a way to create arguments, for example:

```ruby
transform do |r|
  method_call(MyCSVSource, filename: r)
end
```

This `method_call` can be defined simply as:

```ruby
ruby2_keywords def method_call(*args)
  args
end
```

This means an incompatibility for your DSL, so I'm unsure if you can accept this or not.  I'll try to think of another solution.

Sorry again for bothering you, and sorry for my super-long comment!

**Comment by thbar** [comment]
Woah thank you @mame - don't be sorry, it is good to see Ruby making good progress toward something much better!

I deeply appreciate all you have provided to help my humble gem ^_^, this is not a bother at all!

I will play further with all this on real-life projects, and will see what must be done ultimately.

Again, thanks a ton for your in-depth feedback, very much appreciated!

**Comment by thbar** [comment]
I've ultimately decided to deprecate `SourceTransformAdapter`, and provide an upgrade path relying on Ruby's `Enumerator` instead. This lets the user instantiate their class as needed, without any assumptions, which is ultimately probably clearer.

CSV destination and source have been fixed for new keyword arguments.

See Changes.md for details.

Merging to publish the next release.

**Pull request merged**